src/Symfony/Component/Form/FormRegistry.php

```diff
@@ -102,22 +105,41 @@ public function getType($name)
      */
     private function resolveType(FormTypeInterface $type)
     {
-        $typeExtensions = array();
-        $parentType = $type->getParent();
-        $fqcn = get_class($type);
-
-        foreach ($this->extensions as $extension) {
-            $typeExtensions = array_merge(
-                $typeExtensions,
-                $extension->getTypeExtensions($fqcn)
-            );
-        }
+        try {
+            $typeExtensions = array();
+            $parentType = $type->getParent();
+            $fqcn = get_class($type);
 
-        return $this->resolvedTypeFactory->createResolvedType(
-            $type,
-            $typeExtensions,
-            $parentType ? $this->getType($parentType) : null
-        );
+            if (isset($this->checkedTypes[$fqcn])) {
+                $types = implode(' > ', array_merge(array_keys($this->checkedTypes), array($fqcn)));
+                throw new LogicException(sprintf('Circular reference detected for form "%s" (%s).', $fqcn, $types));
+            }
+
+            $this->checkedTypes[$fqcn] = true;
+
+            if ($parentType === $fqcn) {
+                throw new LogicException(sprintf('Form "%s" cannot have itself as a parent.', $fqcn));
+            }
+
+            foreach ($this->extensions as $extension) {
+                $typeExtensions = array_merge(
+                    $typeExtensions,
+                    $extension->getTypeExtensions($fqcn)
+                );
+            }
+
+            try {
+                return $this->resolvedTypeFactory->createResolvedType(
+                    $type,
+                    $typeExtensions,
+                    $parentType ? $this->getType($parentType) : null
+                );
+            } finally {
+                unset($this->checkedTypes[$fqcn]);
+            }
+        } finally {
+            $this->checkedTypes = array();
```

Comment:
I think this is buggy: there should be only one try/finally, it should start right after `$this->checkedTypes[$fqcn] = true;`, and should only `unset($this->checkedTypes[$fqcn]);` (doing `$this->checkedTypes = array()` could be possible, but only on exceptions - but is not required if we unset properly)